### PR TITLE
#169257399 Builds feature Reset password via email

### PR DESCRIPTION
**Pull request by b0nbon1**
#### What does this PR do?
 reset the users' passwords and notifies the users about the changes.
#### Description of Task to be completed?
- [x] send user Reset email with a link
- [x] use a third party API To send mail
- [x] add reset user password template
- [x] send user success reset password
#### How should this be manually tested?
- git clone the repository
- run `npm install` to install dependencies
- get a Sendgrid API key and add it to .env(its sample is `.env.expample` file on the root folder)
##### option 1
- go to the postman and register a new user on endpoint `localhost:3000/api/v1/auth/signup`
- run endpoint `localhost:3000/api/v1/auth/forgotPassword` with the email of the registered user so as to request for a link to update password to your mail
- the check your email it has the link to reset the password
- Click the link and then copy it
-  copy it and run it on postman with body `{"password" : "newpasswordnjd38" }` and use `PATCH` method to run it. 
##### option 2
- run `npm pretest-travis` to migrate to the test tables
- run `npm test` to run the tests.
#### Any background context you want to provide?
- no Interface implemented to handle the reset password so you have to copy the link and test it on postman.
- [Implementation plan ](https://docs.google.com/document/d/1Hd14pPJW6BdoDfb2no4OonPY05WNjZ_adUareJO66bg/edit?usp=sharing)
#### What are the relevant pivotal tracker stories?
[#169257399 ](https://www.pivotaltracker.com/n/projects/2407417/stories/169257399)

### CODE REVIEW

File: src/utils/error.js [hunk 1]
```diff
@@ -4,14 +4,12 @@
...
-     */
-  constructor(message, statusCode = 500) {
-    super();
-    this.statusCode = statusCode;
-    this.message = message;
+   * use to wrap around controller to avoid try catch blocks
+   * @param {function} fn
+   * @returns {function} next
+   */
+  static catchErrors(fn) {
```
Comment (RaymondMwaura) [inline src/utils/error.js]: I think we should avoid classes and static methods and defer to objects:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Details_of_the_Object_Model

File: src/middlewares/checkToken.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * function - decodes the token from the email verification
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ * @returns {undefined}
+ */
+const checkToken = async (req, res, next) => {
+  const { token } = req.query;
```
Comment (RaymondMwaura) [inline src/middlewares/checkToken.js:13]: This is okay for this particular endpoint but for other routes, I think we should pass the token in the 'Bearer Token Authorization Header'.

File: src/services/Mailer.services.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import pug from 'pug';
+import sgMail from '@sendgrid/mail';
+import config from '../config';
+
+sgMail.setApiKey(config.SENDGRID_API_KEY);
+
+/**
+* Class representing a mailing services.
+*/
+export default class mailer {
```
Comment (RaymondMwaura) [inline src/services/Mailer.services.js]: I think we should avoid classes and static methods and defer to objects:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Details_of_the_Object_Model

File: src/services/User.service.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import db from '../models';
+
+/**
+ * Class User services creates a middleware
+ */
+class UserServices {
```
Comment (RaymondMwaura) [inline src/services/User.service.js:6]: I think we should avoid classes and static methods and defer to objects:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Details_of_the_Object_Model

**Review by RaymondMwaura — CHANGES_REQUESTED** [review]
Unit and integration tests are passing. I successfully received the reset password link when I used the forgot-password endpoint.

The validation is not working properly on this route though. Postman returns the following error if a user doesn’t give an email in the request:
```
{
    "status": "error",
    "message": "WHERE parameter \"email\" has invalid \"undefined\" value - check logs for more info"
}
```

The reset password link didn’t work for me, I get the error:
```
status
"error"
message
"Not found"
```

You need to edit your PR test description. The URL to test the feature is not:
localhost:3000/api/v1/forgotpassword
But actually:
localhost:3000/api/v1/auth/forgotpassword

Also, change this message as follows:
Successful reset password please check your email -> Successful password reset, please check your email

File: src/services/Mailer.services.js (hunk 3 above)
Comment (b0nbon1) [reply src/services/Mailer.services.js]: I'm using classes because I need a constructor there. So as to pass data to non-static classes

**Review by RaymondMwaura — DISMISSED** [review]
This is well implemented and all the tests are working. Maybe you can update your test description on your PR to show the user that they should make a 'PATCH' request on Postman using the reset password link received from the email.

**Comment by b0nbon1** [comment]
> Unit and integration tests are passing. I successfully received the reset password link when I used the forgot-password endpoint.
> 
> The validation is not working properly on this route though. Postman returns the following error if a user doesn’t give an email in the request:
> 
> ```
> {
>     "status": "error",
>     "message": "WHERE parameter \"email\" has invalid \"undefined\" value - check logs for more info"
> }
> ```
> 
> The reset password link didn’t work for me, I get the error:
> 
> ```
> status
> "error"
> message
> "Not found"
> ```
> 
> You need to edit your PR test description. The URL to test the feature is not:
> localhost:3000/api/v1/forgotpassword
> But actually:
> localhost:3000/api/v1/auth/forgotpassword
> 
> Also, change this message as follows:
> Successful reset password please check your email -> Successful password reset, please check your email

@RaymondMwaura I think we solved that. Thanks for noting that

**Comment by b0nbon1** [comment]
@RaymondMwaura Its `PATCH` method because I'm just updating a single entity of model users which makes partial `PUT` method is used to update everything.

**Review by gildniy — DISMISSED** [review]
Testing this, following the whole process, from the Password request link up to the New password generation, it seemed to work well. I think it's ok to be merged.

**Review by nshaphan — DISMISSED** [review]

File: package.json [hunk 5]
```diff
@@ -13,13 +13,15 @@
   },
   "scripts": {
     "test": "NODE_ENV=test nyc mocha -r dotenv/config  --require @babel/register \"./src/tests/**/*.js\"   --timeout 150000 --exit",
-    "pretest": "NODE_ENV=test npm run migrate:undo && NODE_ENV=test  npm run migrate",
+    "pretest-Travis": "NODE_ENV=test npm run migrate:undo && NODE_ENV=test  npm run migrate",
```
Comment (timolinn) [inline package.json]: use lowercase characters.

File: src/controllers/users.js [hunk 6]
```diff
@@ -118,29 +107,88 @@ export default class {
...
+    );
+  }
+
+  /**
+   * forgot password controller
+   * @param {*} req
+   * @param {*} res
+   * @returns {*} responses
+   */
+  static async forgotPassword(req, res) {
```
Comment (timolinn) [inline src/controllers/users.js]: use instance method.

File: src/controllers/users.js [hunk 7]
```diff
@@ -118,29 +107,88 @@ export default class {
    * @param {function} next
    * @returns {object} responses
    */
-  static async resendEmail(req, res, next) {
-    try {
-      const { email } = req.query;
-      const hostUrl = `${req.protocol}://${req.get('host')}`;
-      const user = await db.user.findOne({
-        where: { email }
+  static async resendEmail(req, res) {
```
Comment (timolinn) [inline src/controllers/users.js]: use instance method.

File: src/services/Mailer.services.js [hunk 8]
```diff
@@ -5,17 +5,25 @@ import config from '../config';
 sgMail.setApiKey(config.SENDGRID_API_KEY);
 
 /**
-* Class representing a .
+* Class representing a mailing services.
 */
 export default class mailer {
```
Comment (timolinn) [inline src/services/Mailer.services.js]: Change class name to `Mailer`

File: src/services/Mailer.services.js [hunk 9]
```diff
@@ -5,17 +5,25 @@ import config from '../config';
...
+  constructor(data) {
+    this.data = data;
+  }
+
   /**
    * Create a point.
    * @param {object} mailOptionsObject - contains the details to be sent
    * @return {object} status - it has sent of the sent email
    */
   static async sendEmail(mailOptionsObject) {
```
Comment (timolinn) [inline src/services/Mailer.services.js]: use instance method.

File: src/utils/error.js [hunk 10]
```diff
@@ -1,18 +1,9 @@
+
 /**
- * Class representing errorHandlers.
- * @extends Error
- */
-class ErrorHandler extends Error {
```
Comment (timolinn) [inline src/utils/error.js:5]: Why remove this?

**Review by timolinn — CHANGES_REQUESTED** [review]

File: src/utils/error.js (hunk 10 above)
Comment (b0nbon1) [reply src/utils/error.js:5]: wow, didn't notice I removed. Brought it back now

**Comment by timolinn** [comment]
The endpoints specified in the PR description seems is incorrect. I think it should be `localhost:3000/api/v1/auth/forgotPassword` same for the signup endpoint.

**Comment by b0nbon1** [comment]
@timolinn I have noticed the mistake. Thanks. done with the updates

**Review by gildniy — APPROVED** [review]
Tested and working

File: src/controllers/users.js [hunk 11]
```diff
@@ -117,29 +106,88 @@ export default class {
...
+   * forgot password controller
+   * @param {*} req
+   * @param {*} res
+   * @returns {*} responses
+   */
+  async forgotPassword(req, res) {
+    const host = `${req.protocol}://${req.get('host')}`;
+    const user = await UserServices.findUserByEmail(req.body.email);
+    if (!user) {
+      return Responses.handleError(404, 'User with such email does not exist', res);
```
Comment (timolinn) [inline src/controllers/users.js:147]: For security reasons, I think it'd be a great idea to not tell the user if the email exist on our system or not. That way a malicious user cannot guess somebody's email. However, we can go with this, just pointing it out for future purposes.

**Review by timolinn — APPROVED** [review]

**Comment by timolinn** [comment]
Works great!

**Review by nshaphan — APPROVED** [review]
Great work, this work as expected

**Pull request merged**